# Тема 9. Введение в функциональное программирование

# Семинар 9.1 (21)

## 1. Особенности функционального програмирования

В ходе развития декларативной парадигмы сформировался функциональный подход, в котором процесс вычисления трактуется как вычисление значений функций в их математическом понимании (иногда используется термин «чистая функция»). «Чистая» функция зависит только от ее входных данных и результатов работы других функций, а не от внешнего контекста. Таким образом, явное хранение или изменение состояния программы не предусмотрено, поэтому функциональному программированию, как правило, присущи такие свойства, как:

- отсутствие глобальных и локальных переменных;
- неизменность переменных, т. е. переменная после инициализации не изменяется, но можно на ее основе создать другую переменную с новым значением;
- функции высшего порядка, когда функция может принимать другую функцию в качестве своего аргумента или возвращать другую функцию в качестве результата;
- представление аргументов и результатов работы каждой функции в виде списков.

Одним из самых популярных и наглядных элементов функциональной парадигмы является механизм рекурсии. При рекурсивной реализации для получения результата функция может многократно вызывать саму себя с новым набором аргументов до выполнения определенного стоп-условия, которое позволяет выдать готовый ответ для какого-либо тривиального значения аргумента.

<u>Пример 1</u>

In [14]:
# Поиск наибольшего общего делителя в импертивном стиле
def gcd_imp(a, b):
    r = 0
    while b > 0:
        r = a % b
        print(f'a: {a}, b: {b}, r: {r}')
        a, b = b, r
    return a

In [15]:
gcd_imp(9702, 945)

a: 9702, b: 945, r: 252
a: 945, b: 252, r: 189
a: 252, b: 189, r: 63
a: 189, b: 63, r: 0


63

In [17]:
# Поиск наибольшего общего делителя в функциональном стиле
def gcd_func(a, b):
    print(f'a: {a}, b: {b}')
    if b == 0:
        return a
    else:
        return gcd_func(b, a%b)

gcd_func(9702, 945)

a: 9702, b: 945
a: 945, b: 252
a: 252, b: 189
a: 189, b: 63
a: 63, b: 0


63

## 2. Основы: функции первого класса, функции высшего порядка, замыкание

Неотъемлемой частью функционального программирования является возможность передачи функций в качестве аргументов другим функциям, возврат их как результат других функций, присваивание их переменным или сохранение в структурах данных. То есть основная идея заключается в том, что функции (или методы, если речь одет об ООП) имеют тот же статус, что и другие объекты данных. К слову, функции, которые могут передаваться в качестве аргументов, называют ***функциями первого класса***, а функции, которые их могут принимать в качестве аргументов (или возвращать в качестве результата) — ***функциями высшего порядка***.

<u>Пример 2</u>

In [1]:
def square(x):
    return x**2

def cube(x):
    return x**3

def my_map(func, arg_list):
    result = []
    for i in arg_list:
        result.append(func(i))
    return result

sq = my_map(square, [1, 2, 3, 4])
print(sq)

cb = my_map(cube, [1, 2, 3, 4])
print(cb)

[1, 4, 9, 16]
[1, 8, 27, 64]


Приведен пример использования функций *square()* и *cube()* в качестве аргумента при вызове другой функции — *my_map()*. Здесь важно понимать, что в функцию *my_map()* в качестве аргумента передается не результаты выполнения функции *square()* или *cube()*, а именно сама функция (точнее, указатель на нее). Это позволяет вызывать нужную функцию из тела функции *my_map()*.

<u>Пример 3</u>

In [2]:
def set_func(value):
    base = value
    
    def add_func(shift):
        temp = base + shift
        return temp
    
    return add_func #тут возвращается не результат, а адрес функции add_func

f1 = set_func(10)
print(f1)
print(f1(2))

f2 = set_func(100)
print(f2)
print(f2(22))

print(f1(14))

<function set_func.<locals>.add_func at 0x000002293A4B2160>
12
<function set_func.<locals>.add_func at 0x000002293A4B2200>
122
24


В этом примере внутри функции *set_func()* описана вложенная функция — *add_func()*. В то же время нигде не видно, как вызывается эта функция: оператор `return add_func` в конце функции *set_func()* не вызывает функцию *add_func()* — это следует из отсутствия скобок после имени add_func, а возвращает саму функцию (точнее, ее адрес). Поэтому результатом работы функции *set_func()* является адрес функции *add_func()*, т. е. в данном примере функция является возвращаемым значением.

***Замыкание*** — это функция первого класса, в теле которой присутствуют ссылки на переменные, объявленные вне тела этой функции в окружающем коде и не являющиеся её параметрами. Говоря другим языком, замыкание — функция, которая ссылается на свободные переменные в своей области видимости.

Функция замыкания обязательно определена внутри тела другой функции и создаётся каждый раз во время её выполнения. При этом внутренняя функция содержит ссылки на локальные переменные внешней функции. Каждый раз при выполнении внешней функции происходит, во-первых, создание нового экземпляра области видимости и, во-вторых, создание нового экземпляра внутренней функции, с новыми ссылками на переменные внешней функции. Важно отметить, что ссылки на переменные внешней функции действительны внутри вложенной функции до тех пор, пока работает вложенная функция, даже если внешняя функция закончила работу, и переменные вышли из области видимости — в этом и состоит суть замыкания.

Выполнение оператора `f1 = set_func(10)` приведет к тому, что будет создана область видимости в пределах функции *set_func()* и в переменную *f1* будет записан адрес функции *add_func()*. Поэтому вызов оператора *print()* для *f1* приведет к выводу информации об этой функции — видно, что она является локальной функцией для функции *set_func()*.

Поскольку в переменную f1 записан адрес функции, то можно вызвать эту переменную как обычную функцию, указав в скобках необходимый параметр вызова — в данном случае число 2. В результате этого шага будет вызван экземпляр функции add_func(), который соответствует области видимости со значением переменной *base*, равным $10$. Поэтому результатом вызова будет результат операции $10+2$. Благодаря замыканию объекты из окружающей области видимости записываются и сохраняются после того, как функция (в данном случае *set_func()*) закончила свое выполнение. Эти объекты защищены от сборщика мусора.

В следующей строке вызывается функция *set_func()* с другим аргументом и записывается результат вызова в переменную f2. Очевидно, что записанное в f2 значение будет совпадать со значением в f1 (и там, и там будет содержаться адрес функции add_func()). Однако при новом вызове *set_func()* будет создан новый экземпляр области видимости, в котором значение *base* будет равно $100$. Поэтому теперь, при косвенном вызове функции *add_func()* через переменную *f2* получим в результате значение $100 + 22 = 122$.

Обратите, внимание, что область видимости, связанная с переменной *f1* сохранилась. Если мы захотим снова вызвать функцию *add_func()* через переменную *f1*, то получим доступ к первому экземпляру области видимости, где *base* равно $10$.

Замыкание, так же, как и экземпляр объекта, есть способ представления функциональности и данных, связанных и упакованных вместе, только здесь инкапсуляция данных и методов для их обработки осуществляется не на уровне классов (как в объектно-ориентированном программировании), а на уровне функций (функциональное программирование).

## 3. Декораторы

Декоратор — это функция, которая принимает функцию или метод в качестве единственного аргумента и возвращает новую функцию или метод, включающую декорированную функцию или метод, с дополнительными функциональными возможностями.

Существует два виде декорирования функции, зависящих от того, есть ли у декорируемой функции входные параметры или нет.


## 3.1. Декорирование функции без параметров

Рассмотрим процесс декорирования функции без параметров. Представим процесс декорирования функции *my_func*, которая в теле функции не содержит *return*.

<u>Пример 4</u>

In [5]:
def my_func():
    print('Работа функции my_func')

my_func()

Работа функции my_func


In [6]:
def func_decorator(func):
    def wrapper():
        print('_ _ _ Действия перед вызовом функции my_func _ _ _')
        func()
        print('_ _ _ Действия после выполнения функции my_func _ _ _')
    return wrapper

my_func = func_decorator(my_func)
my_func()

_ _ _ Действия перед вызовом функции my_func _ _ _
Работа функции my_func
_ _ _ Действия после выполнения функции my_func _ _ _


Декоратор очень похож на замыкание, но у него есть две особенности:

1) В качестве параметра внешняя функция обязательно получает адрес какой-либо «третьей» функции, функционирование которой, собственно, и надо изменить с помощью декоратора. В примере 3 в роли такой декорируемой функции выступает функция *my_func()* с единственным параметром вызова. Первый вызов оператора *print()* показывает, что эта функция является обычным объектом программы, не связанным с какими-либо другими функциями. При вызове функции *my_func()* на консоль выводится константная строка «Работа функции my_func».

2) Далее, результат вызова внешней функции *func_decorator()* записывается не просто в какую-нибудь произвольную переменную, а в переменную, имя которой совпадает с именем той функции, поведение которой необходимо изменить (т. е. декорируемой функции). В примере 3 в качестве такой переменной выступает *my_func*. Получаем, что вначале программы это имя связано с функцией *my_func()*, а в конце программы — с внутренней функцией *wrapper()*. Теперь, при вызове в программе функции *my_func()* сначала будет осуществляться переход на функцию *wrapper()*, затем вывод строки «Действия перед вызовом функции my_func». После этого из вызванной функции wrapper() вызывается c одним параметром функция *my_func()*, в которой на консоль выводится «Работа функции my_func» и выполняется вывод строки «Действия после выполнения функции my_func».

Таким образом, получается, что при помощи декоратора изменили поведение функции, имя которой записано в переменную *my_func ()*. В Python предусмотрена возможность использования декораторов в коде с применением специального синтаксиса с символом @ (коммерческое at). Представим процесс декорирования функции *my_func*, которая возвращает значение в основной программный код.

<u>Пример 5</u>

In [12]:
def func_decorator(func):
    def wrapper():
        print('_ _ _ Действия перед вызовом функции my_func _ _ _')
        res = func('test')
        print('_ _ _ Действия полсле выполения функции my_func _ _ _')
        return res
    return wrapper

@func_decorator
def my_print(temp):
    return f'Работа функции my_print, {temp}'

my_print()

_ _ _ Действия перед вызовом функции my_func _ _ _
_ _ _ Действия полсле выполения функции my_func _ _ _


'Работа функции my_print, test'

Синтаксис описания функции-декоратора с использованием *@* повышает читаемость кода, однако по сути является ничем иным, как «синтаксическим сахаром», без которого вполне можно обойтись.


## 3.2. Декорирование функции с произвольным числом параметров

Рассмотрим еще пример применения декоратора, в котором в котором одну функцию *decor_func()* необходимо применить для двух разных функций: *display()* и *display_info()*. Вариант с *display()* работает корректно, а вот вариант с *display_info()* выдает ошибку. Это связано с тем, что в соответствие с логикой работы декорируемой функции она должна брать параметры из своего локального окружения (области видимости). Однако при создании декоратора *decor_func()* эти параметры не были переданы в область видимости, так как у функции *wrapper()* нет аргументов и она не может их передать в функцию *orig_f()* при вызове последней.

<u>Пример 6</u>

In [13]:
def decor_func(orig_f):
    def wrapper():
        print(f'Этот код выполняется перед вызовом {orig_f.__name__}')
        return orig_f()
    return wrapper

@decor_func
def display():
    print('Выполняется display')

@decor_func
def display_info(name, age):
    print(f'Выполняется display_info с аргументами ({name}, {age})')

try:
    display()
    print('Успешно')
except TypeError as e:
    print('Ошибка выполнения display()')
    print(e)

print()
try:
    display_info('Иван', 25)
    print('Успешно')
except TypeError as e:
    print('Ошибка выполнения display_info()')
    print(e)

Этот код выполняется перед вызовом display
Выполняется display
Успешно

Ошибка выполнения display_info()
decor_func.<locals>.wrapper() takes 0 positional arguments but 2 were given


Следовательно, если необходимо через один и тот же декоратор выполнять вызовы разных функций, то у всех этих функций должны быть одинаковые списки параметров вызова и эти аргументы должны «пробрасываться» через декорируемую функцию.

Существует универсальный вариант разработки декоратора для нескольких функций с различными входными параметрами. В Python можно передать переменное количество аргументов двумя способами:

- в виде списка неименованных (или еще говорят «позиционных» аргументов) — в этом случае они помещаются в кортеж;
- в виде списка именованных аргументов — в этом случае они помещаются в словарь.

Аргументами функции *wrapper()* стали объекты с именами \*args и \*\*kwargs. Аргумент \*args собирает позиционные аргументы, а \*\*kwargs — именованные. Например, в вызове `wrapper(1, 'a', x=5, y=None)` значение args — это кортеж `(1, 'a')`, а kwargs — это словарь `{'x': 5, 'y': None}`. Если позиционных аргументов при вызове функции нет, args — пустой; если  именованных аргументов нет, kwargs — пустой.

<u>Пример 7</u>

In [8]:
def decor_func(orig_f):
    def wrapper(*args, **kwargs):
        print(f'Этот код выполняется перед вызовом {orig_f.__name__}')
        print(args)
        return orig_f(*args, **kwargs)
    return wrapper

@decor_func
def display():
    print(f'Выполняется display с аргументами ')

@decor_func
def display_info(name, age):
    print(f'Выполняется display_info с аргументами ({name}, {age})')

display()
display_info('Иван', 25)

Этот код выполняется перед вызовом display
()
Выполняется display с аргументами 
Этот код выполняется перед вызовом display_info
('Иван', 25)
Выполняется display_info с аргументами (Иван, 25)


Благодаря такому изменению способа «пробрасывания» параметров, вызываемая через декоратор функция может иметь произвольное число аргументов, то есть можно использовать один и тот же декоратор для функций с разным числом аргументов.

<u>Пример 8</u>

In [10]:
def my_decorator(func):
    def wrapper(*args, **kwargs):
        print('_ _ _ Действия перед вызовом функции my_func _ _ _')
        func(*args, **kwargs)
        print('_ _ _ Действия после выполнения функции my_func _ _ _')
    return wrapper

@my_decorator
def my_print(param):
    print(f'Функция выводит {param}')

my_print('Привет, Мир!')

_ _ _ Действия перед вызовом функции my_func _ _ _
Функция выводит Привет, Мир!
_ _ _ Действия после выполнения функции my_func _ _ _


## Декоратор @property и функция property()

В Python есть несколько встроенных декораторов, также существуют декораторы в различных разделах стандартной библиотеки Python. С примерами встроенных декораторов, такими как @staticmethod и @classmethod мы уже знакомы. Теперь давайте познакомимся с одним из наиболее часто используемых в ООП декоратором — @property.

Декоратор @property перед методом класса говорит о том, что создается атрибут для чтения с таким же именем и этот метод возвращает его значение. То есть вместо работы с методом класса получаем возможность работы с атрибутом с тем же именем как с полем данных. Таким образом организуется доступ к внутреннему состоянию объекта, имитирующий обращение к его полю, правда только для чтения.

В примере 8 в последней строке кода мы можем обратиться к методу *area* экземпляра класса *Rectangle* как к полю данных. То есть не нужно вызывать метод *area()*. Вместо этого при обращении к *area* как к атрибуту (то есть без использования скобок), соответствующий метод вызывается неявным образом. Это возможно благодаря декоратору @property, который работает поверх встроенной функции property().

Размещение конструкции @property перед определением функции равносильно использованию конструкции вида `area = property(area)`.

<u>Пример 9</u>

In [4]:
class Rectangle(object):
    def __init__(self, a, b):
        self.a = a
        self.b = b
    
    @property
    def area(self):
        return self.a * self.b

rect = Rectangle(5, 6)
rect.area

30

Функция *property()* — интересная встроенная функция языка Python, которая принимает одну или несколько других функций в качестве аргумента и использует их вместо обращения к полю данных для чтения, записи или удаления — так, как это показано в примере 9. Благодаря функции *property()* можно оставить методы *getColor()* и *setColor()* приватными методами класса *Pen* и не открывать к ним доступ для сторонних объектов. А вместо этого использовать доступ к искусственному полю *color* на чтение или на запись для вызова этих методов

<u>Пример 10</u>

In [5]:
class Pen(object):
    def __init__(self, new_color):
        self.__color = new_color
    
    def __getColor(self):
        return (self.__color)
    
    def __setColor(self, new_color):
        self.__color = new_color
    
    color = property(__getColor, __setColor)

pen1 = Pen(5)
print(pen1.color)
pen1.color = 22
print(pen1.color)

5
22


### Задание 1

Создайте декоратор, который будет выводить на экран аргументы функции и их типы.

In [25]:
# 1

def my_deco(func):
    def wrapper(*args):
        print([(x, type(x)) for x in args])
    return wrapper

@my_deco
def func1():
    pass

func1(1, 2, 3, 'doll', 'Barbie')

[(1, <class 'int'>), (2, <class 'int'>), (3, <class 'int'>), ('doll', <class 'str'>), ('Barbie', <class 'str'>)]


### Задание 2

Создайте декоратор, который будет выводить время выполнения функции и сохранять его в файл.

In [22]:
# 2
from datetime import datetime
import time

def time_deco(func):
    def wrapper():
        start = datetime.now().second
        func()
        finish = datetime.now().second
        
        with open('Result.txt', 'w') as f:
            f.write(f'Время работы: {finish - start} сек.')
        
        return finish - start
    return wrapper
    
@time_deco
def func1():
    time.sleep(3)
    print('Работает функция')

work_time = func1()

Работает функция


### Задание 3

Создайте декоратор, который будет выводить на экран статистику вызовов функции и сохранять ее в базе данных.

In [21]:
import datetime

callFunc = {}

def time_call(func):
    def wrapper(*args, **kwargs):
        print('Тут начало функции')
        print(f'Вызов функции произошел в {datetime.datetime.now()}')
        callFunc[datetime.datetime.now()] = func.__name__
        func(*args, **kwargs)
        print('Тут конец функции\n')
    return wrapper

@time_call
def test_func(test):
    print(test)
    
    
test_func(1)
test_func(2)
test_func(3)
test_func(4)
test_func(5)

print(callFunc)

Тут начало функции
Вызов функции произошел в 2024-03-15 15:12:45.500377
1
Тут конец функции

Тут начало функции
Вызов функции произошел в 2024-03-15 15:12:45.500377
2
Тут конец функции

Тут начало функции
Вызов функции произошел в 2024-03-15 15:12:45.500377
3
Тут конец функции

Тут начало функции
Вызов функции произошел в 2024-03-15 15:12:45.500377
4
Тут конец функции

Тут начало функции
Вызов функции произошел в 2024-03-15 15:12:45.500377
5
Тут конец функции

{datetime.datetime(2024, 3, 15, 15, 12, 45, 500377): 'test_func'}
